### ISSUE TITLE
Add __reduce__ function to allow pickling of pynfft object

### ISSUE THREAD

**Issue by chaithyagr** [pyNFFT/pyNFFT#71]
Most parallel processes usage work by pickling the objects and unpickling in each process where it can create a new object per process. However, currently, we cannot run pynfft in parallel because:

Currently, the pyNFFT object cannot be pickled:

```python
import pynfft
import pickle
plan = pynfft.NFFT(N=(512,512), M=1000)
pickle.dump(plan, open('test.pkl', 'wb'))
```

We obtain the following error:

`TypeError: no default __reduce__ due to non-trivial __cinit__`

Based on python guidelines: https://docs.python.org/3.6/library/pickle.html#object.__reduce__, we need to set up a __reduce__ function that instructs on how to pickle the object so that it can be created again on loading.

**Comment by kohr-h** [comment]
Hi @chaithyagr. Thanks for your proposal! Unfortunately, I think serializing a plan wouldn't really be a good idea, mainly because a plan by construction owns the input and output arrays and has some other internal state in the C struct that is not exposed to Python. Properly serializing and transmitting a plan would be non-trivial to implement and potentially result in very costly operations because you'd have to transmit the input and output arrays alongside.

I think it's better to create the plan anew where you need it, not send it over the wire.

**Comment by chaithyagr** [comment]
Thank you for your reply @kohr-h .

> I think it's better to create the plan anew where you need it, not send it over the wire.

This is what we do currently in our implementations.

Would you say that the cost of creating a plan is less than getting a loaded plan from memory?

While we try to use parallel processes and create a plan on-site, we will have to make N different plans, rather we could create 1 and copy to others right?

I do understand the implementation is non-trivial, I tried some changes and failed.

**Comment by kohr-h** [comment]
> Would you say that the cost of creating a plan is less than getting a loaded plan from memory?

> While we try to use parallel processes and create a plan on-site, we will have to make N different plans, rather we could create 1 and copy to others right?

The cost could be significant, in particular due to the pre-computation stuff going on. But loading a plan from memory would mean to recreate all the internal state of the NFFT plan, whose details are deep down in the C library. There are no public functions I know of to, e.g., load the pre-computations from a file or a bytestring. So short of doing really hacky stuff, I don't see a way to achieve this copying of plans you intend to do.

I guess the best for you caser would be to keep the workers alive and reuse the plans from earlier runs, provided that the sizes don't change. (Or, if you have a few distinct sizes, you could still keep those few plans around.)

**Comment by chaithyagr** [comment]
Thank you @kohr-h 
I understand your point. At this point, I dont know whether to close this issue as will not fix, or keep it open for future in case theres a new library or framework that could allow or make things easier to do this.

**Comment by kohr-h** [comment]
Well, if you're really determined you could try to convince the NFFT devs (https://github.com/NFFT/nfft) to expose functionality for storing plan state and initializing from file or similar. If yes, then we can reconsider on our side. Until then, I'd indeed prefer to "won't fix" close the issue.
(And since we're a Python wrapper of **the** NFFT library, switching to another library is not an option.)

**Comment by kohr-h** [comment]
I'll close this issue. If something changes that makes this doable, feel free to re-open it @chaithyagr.
